lisa/tools/wsl.py

```diff
@@ -0,0 +1,332 @@
+            done = False
+            while elapsed.elapsed(False) < self.INSTALL_TIMEOUT:
+                if self._check_install_done(distro=name):
+                    done = True
+                    break
+                time.sleep(1)
+
+            # raise error if not done
+            if not done:
+                self._check_install_done(distro=name, raise_error=True)
+
+            self.shutdown(name)
+
+        self.reload_guest_os()
+
+        if not is_installed:
+            # set NOPASSWD
+            echo = self._guest.tools[Echo]
+            echo.write_to_file(
+                "sudo ALL=(ALL:ALL) NOPASSWD:ALL",
+                self._guest.get_pure_path("/etc/sudoers"),
+                append=True,
+            )
+
+    def normalize_result(self, result: ExecutableResult) -> ExecutableResult:
+        # wsl output is utf-16-le, but Windows returns utf-8. The logic is to
+        # try best to normalize, but still possible not to be normalized. So
+        # calling this method to normalize output explicitly.
+        encoding = charset_normalizer.detect(result.stdout.encode())["encoding"]
+        if encoding != "utf-8":
+            assert isinstance(encoding, str), f"actual {type(encoding)}"
+            result.stdout = result.stdout.encode(encoding).decode()
+            result.stderr = result.stderr.encode(encoding).decode()
+
+        return result
+
+    def shutdown(self, distro: str) -> None:
+        self._log.debug(f"shutting down {distro}")
+        self._wsl_execute(f"--terminate {distro}")
+
+    def reload_guest_os(self) -> None:
+        from lisa.operating_system import OperatingSystem
+
+        self._guest.os = OperatingSystem.create(self._guest)
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        super()._initialize(*args, **kwargs)
+        if not self.node.os.is_windows:
+            raise LisaException("wsl is only available on Windows")
+
+    def _check_exists(self) -> bool:
+        self._detect_installed_path()
+
+        # detecting version, and check if current version support --version
+        # parameter. The version command may not work, use utf-8 to see Windows
+        # error.
+        result = self._wsl_execute("--version", encoding="utf-8")
+
+        return result.exit_code == 0
+
+    def _detect_installed_path(self) -> bool:
+        is_found = False
+
+        if not self.node.is_remote:
+            # local doesn't have OPENSSH issue, use default path.
+            self._command_path = "wsl"
+            return True
+
+        fixed_wsl_path = self.node.expand_env_path(self.FIXED_WSL_PATH)
+        self._log.debug("try to detect wsl path: %s", fixed_wsl_path)
+        exists, _ = self.command_exists(fixed_wsl_path)
+        if exists:
+            # quote the path, because it may have space.
+            self._command_path = f'"{fixed_wsl_path}"'
+            is_found = True
+            self._log.debug("detected wsl path: %s", self._command_path)
```

Comment:
self._log.debug(f"detected wsl path: {self._command_path}")

Reply:
Good catch. It should be generated by co-pilot, haven't noticed it haven't used f string.